src/wincng.h

```diff
@@ -385,7 +394,17 @@ _libssh2_bn *_libssh2_wincng_bignum_init(void);
  * Windows CNG backend: Diffie-Hellman support
  */
 
-#define _libssh2_dh_ctx struct _libssh2_wincng_bignum *
+typedef struct {
+#if LIBSSH2_USE_BCRYPT_DH
+    /* holds our private+public key components */
+    BCRYPT_KEY_HANDLE dh_handle;
+    /* records the parsed out modulus and generator parameters that are shared
+    * with the peer */
+    BCRYPT_DH_PARAMETER_HEADER *dh_params;
+#endif
+    /* fallback if the newer DH api doesn't work on this system */
+    struct _libssh2_wincng_bignum *bn;
```

Comment:
Why aren't you falling back to the original define? I think there is no need to nest the bignum struct in another struct.

Reply:
the determination about whether we can use the DH API is based on a runtime test so both approaches need to be compiled into the same binary.